router/src/main/java/org/simple/clinic/router/screen/ScreenRouter.kt

```diff
@@ -78,16 +80,15 @@ class ScreenRouter(
     return BackStackPopCallback(popped)
   }
 
-  fun sendResultAndPop(result: Any) {
-    // TODO: Now that BaseViewGroupKeyChanger inflates a new view before removing
-    // TODO: the older one, this is no longer be required. Remove this after testing.
-    // The name is actually incorrect. It's important to send the result only
-    // after the previous screen is inflated and has registered for results.
-    // But the name is kept in this way to maintain familiarity with how Activity
-    // results are sent -- the result is set before the Activity is finished.
+  fun popWithResult(key: String, result: Any?) {
+    screenResults.put(key, result)
     pop()
+  }
 
-    resultBus.send(result)
+  @Suppress("UNCHECKED_CAST")
+  @CheckResult
+  fun <T> retrieveResult(key: String): T? {
+    return screenResults.get(key) as T?
```

Comment:
Hrm, one thing I don't understand is how do you plan to remove `streamScreenResults()`? Did you intentionally leave it?

Reply:
Yeah, I was planning to remove it, but I saw its being used for sending permission grant results. Once this gets merged in, I'll create a story for changing permission results to the new flow and remove the streamed results.

Reply:
Do you plan to do it right away? It won't be nice to have two different ways of exchanging screen results. 

I can sit with you and come up with a solution instead of deferring it to later because it won't happen anytime soon otherwise.

Reply:
Off the top of my head, I can think of a solution where `ScreenResults` piggybacks on `ScreenResultBus`. Any screen result sent through `ScreenRouter` is available via both `retrieveResult()` and `streamScreenResults()`. The only difference would be that the latter would be an observable infinite stream. 

Any value sent to `ScreenRouter` could be first sent to `ScreenResultBus` and it could later be observed by `ScreenResults` to hold it.

Reply:
> Do you plan to do it right away? It won't be nice to have two different ways of exchanging screen results.

 I can do this in this PR, actually. It will be nice to have it implemented and functional when we merge this in.

> Any value sent to ScreenRouter could be first sent to ScreenResultBus and it could later be observed by ScreenResults to hold it.

I don't want to do this because then we will run into the same problem that we have now. Late subscribers will not receive the screen results, and if we leave the results in memory we will run into the same issues that we'll have with a `BehaviourSubject`, i.e, screens might receive stale results.

I'd rather that the intended consumers use the result once and convert into a stream before forwarding it to the controllers instead of the screen router exposing a stream.

Reply:
Let's discuss this in person tomorrow

Reply:
If I make the changes, can we merge this in today? The rest of the simple ID feature is blocked on this.

Reply:
We can always discuss and figure out a better solution and implement it later.

Reply:
> The rest of the simple ID feature is blocked on this.

This is why I felt we should implement a stopgap solution in simple ID for now because this would have required a good amount of discussion. I'd hate to merge this under pressure :( 

> We can always discuss and figure out a better solution and implement it later.

I don't want to put in a half baked solution because maintenance stories like this never get done unless something is blocked on it.

Reply:
Your call. I'd recommend placing a stopgap solution in favor of properly improving the app's ui navigation, but I can understand if that's too much an efort.

Reply:
Technically, `Flow` itself is a half-baked solution. :P

IMO, this is a better solution than what we have now because this is an imperative solution that will support both synchronous as well as asynchronous view inflation. Adding in a workaround or a hack for simple ID would be more cumbersome than using this.

Reply:
I'd like to get this in without changing how permission results are delivered and finish up the Simple ID v1 flow. Then we can figure out what is a good solution for doing a `startActivityWithResult` equivalent for our use cases.

Reply:
> Technically, Flow itself is a half-baked solution. :P

Why? What we're trying to solve in this PR is entirely independent of Flow. We wrote the results bus along the lines of how Activities exchange data. We would have still encountered this problem even if we weren't using Flow.

> Adding in a workaround or a hack for simple ID would be more cumbersome than using this.

Not really. We're using `SharedPreferences` for showing the `patient-saved` snackbar and it's working just fine. There's nothing cumbersome about it.

Instead of calling `screenRouter.popWithResult()`, you'd call `preference.set()` and `screenRouter.retrieveResult()` will get replaced by `preference.get()` or `preference.asObservable()`. 

> Then we can figure out what is a good solution for doing a startActivityWithResult equivalent for our use cases.

We already have a `startActivityWithResult()` equivalent in place. The current implementation works exactly like it. :p

Reply:
> > Technically, Flow itself is a half-baked solution. :P
> 
> Why? What we're trying to solve in this PR is entirely independent of Flow. We wrote the results bus along the lines of how Activities exchange data. We would have still encountered this problem even if we weren't using Flow.

The reason I am bringing this is up is that Flow is a half-baked solution that doesn't provide a good way to start a screen with results which is, IMO, an essential feature for any navigation library.

> > Adding in a workaround or a hack for simple ID would be more cumbersome than using this.
> 
> Not really. We're using `SharedPreferences` for showing the `patient-saved` snackbar and it's working just fine. There's nothing cumbersome about it.
> 
> Instead of calling `screenRouter.popWithResult()`, you'd call `preference.set()` and `screenRouter.retrieveResult()` will get replaced by `preference.get()` or `preference.asObservable()`.

I don't see any difference between using shared preferences and what this solution is since shared preferences is a key value store on the disk, and what we have here is an in-memory key-value store. We used shared preferences in the patients screen to work around this problem in the exact same way, except that this is in-memory.

> > Then we can figure out what is a good solution for doing a startActivityWithResult equivalent for our use cases.
> 
> We already have a `startActivityWithResult()` equivalent in place. The current implementation works exactly like it. :p

Not really. We use the permission changes in the `PatientsScreen` and the streaming of permission results works fine there. I guess activity results are being delivered at a later point in time after the `PatientsScreen` is inflated in some way.

Reply:
> The reason I am bringing this is up is that Flow is a half-baked solution that doesn't provide a good way to start a screen with results which is, IMO, an essential feature for any navigation library.

Sure. Flow is just a library for managing backstack. That said, we already have an alternative in place and it works exactly like how `Activity.startActivityForResult()` works.

> I don't see any difference between using shared preferences and what this solution is since shared preferences is a key value store on the disk, and what we have here is an in-memory key-value store. 

Awesome. If you don't see any difference then let's go with SharedPreferences and work on the Screen Router APIs properly in the future 😄 

> We use the permission changes in the PatientsScreen and the streaming of permission results works fine there. I guess activity results are being delivered at a later point in time after the PatientsScreen is inflated in some way.

Correct, `PatientsScreen` is already inflated when the user grants permission so it works out fine in that case.

Reply:
On a second thought, we've spent a lot of time discussing this. I don't think we're going to reach an agreement. Let's go ahead with whatever you feel like doing right now. 

Btw, if ViewPager works out without an adapter then that will be the best.